tensorrt_llm/_torch/auto_deploy/transform/library/kvcache.py

```diff
@@ -0,0 +1,241 @@
+
+import operator
+from typing import Dict, Tuple, Type
+
+from pydantic import ConfigDict, Field
+from torch.fx import Graph, GraphModule, Node
+
+from ...custom_ops.attention_interface import AttentionDescriptor, CacheConfig
+from ...models.factory import ModelFactory
+from ...shim.interface import CachedSequenceInterface
+from ...transformations._graph import add_graph_input, canonicalize_graph
+from ...utils.logger import ad_logger
+from ...utils.node_utils import get_all_input_output_nodes, is_op
+from ..interface import BaseTransform, TransformConfig, TransformInfo, TransformRegistry
+
+
+@TransformRegistry.register("update_in_out_nodes")
+class UpdateInOutNodes(BaseTransform):
+    """Modify the graph module by adding new input nodes and canonicalizing the graph.
+
+    The new input nodes correspond to the extra arguments needed for cached and flattened attention.
+
+    Args:
+        egm: The graph module to analyze and modify.
+        cm: Cached sequence interface containing extra argument information.
+    """
+
+    def _apply(
+        self, gm: GraphModule, cm: CachedSequenceInterface, factory: ModelFactory
+    ) -> Tuple[GraphModule, TransformInfo]:
+        # loop through nodes to get input, output, and get_attr nodes
+        input_nodes, output_nodes = get_all_input_output_nodes(gm.graph)
+
+        # we only expect one input node
+        assert len(input_nodes) == 2, "Expected exactly two input nodes (input_ids, position_ids)."
+
+        # NOTE: for now, we wanna make sure we *only* return the final output and no hidden states.
+        # Later on, we can revisit how to support returning hidden states.
+        assert len(output_nodes) == 1, "Expected exactly one output node!"
+        assert len(output_nodes[0].all_input_nodes) == 1, (
+            "Expected to only return final tensor output!"
+        )
+        ad_logger.info(f"Found {len(input_nodes)} input nodes and {len(output_nodes)} output nodes")
+
+        # Activate and add extra argument nodes
+        new_args = cm.info.switch_to_cached_attn_inputs()
+        for name in new_args:
+            input_nodes.append(add_graph_input(gm, name))
+        ad_logger.info(f"Added {len(new_args)} new input nodes for cached attention metadata")
+
+        canonicalize_graph(gm)
+
+        # TODO:(hg) confirm this
+        info = TransformInfo(skipped=False, num_matches=1, is_clean=True, has_valid_shapes=True)
+
+        return gm, info
+
+
+class InsertCachedAttentionConfig(TransformConfig):
+    """Configuration for the insert cached attention transform."""
+
+    # NOTE:type CacheConfig contains a torch.dtype not supported by pydantic. We set this to avoid error.
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    attn_descriptor: Type[AttentionDescriptor] = Field(
+        description="The attention descriptor to use."
+    )
+    cache_config: CacheConfig = Field(description="The cache config to use.")
```

Comment:
cache_config comes from the model factory. So instead of putting it into the config of the transform, you should be able to retrieve from the factory provided to `_apply`